picky/src/ssh/private_key.rs

```diff
@@ -0,0 +1,644 @@
+
+        let base_key = SshBasePrivateKey::Rsa(private_key);
+
+        SshPrivateKey {
+            _cipher_name: cipher_name,
+            kdf,
+            base_key,
+            public_key,
+            check: 0,
+            comment: comment.unwrap_or_default(),
+            passphrase,
+        }
+    }
+}
+
+impl From<PrivateKey> for SshPrivateKey {
+    fn from(private_key: PrivateKey) -> Self {
+        SshPrivateKey::h_picky_private_key_to_ssh_private_key(private_key, None, None)
+    }
+}
+
+impl SshPrivateKeyParser for SshPrivateKey {
+    type Error = SshPrivateKeyError;
+
+    fn decode(mut stream: impl Read, passphrase: Option<String>) -> Result<Self, Self::Error>
+    where
+        Self: Sized,
+    {
+        let mut auth_magic = [0; AUTH_MAGIC.as_bytes().len()];
+        stream.read_exact(&mut auth_magic)?;
+        if auth_magic != AUTH_MAGIC.as_bytes() {
+            return Err(SshPrivateKeyError::InvalidAuthMagicHeader);
+        }
+        stream.read_u8()?; // skip 1 byte (null-byte)
+
+        let cipher_name = stream.read_ssh_string()?;
+        let kdf_name = stream.read_ssh_string()?;
+        let kdf_option: KdfOption = SshComplexTypeDecode::decode(&mut stream)?;
+        let keys_amount = stream.read_u32::<BigEndian>()?;
+
+        if keys_amount != 1 {
+            return Err(SshPrivateKeyError::InvalidKeysAmount(keys_amount));
+        }
+
+        // read public key
+        let _ = stream.read_ssh_bytes()?;
+
+        // read private key
+        let private_key = stream.read_ssh_bytes()?;
+
+        let data = decrypt(&cipher_name, &kdf_name, &kdf_option, passphrase.as_deref(), private_key)?;
+
+        let mut cursor = Cursor::new(data);
+
+        let check0 = cursor.read_u32::<BigEndian>()?;
+        let check1 = cursor.read_u32::<BigEndian>()?;
+        if check0 != check1 {
+            return Err(SshPrivateKeyError::InvalidCheckNumbers(check0, check1));
+        }
+
+        let base_key: SshBasePrivateKey = SshComplexTypeDecode::decode(&mut cursor)?;
+        let base_public_key = base_key.base_public_key();
+
+        let comment = cursor.read_ssh_string()?.trim_end().to_owned();
+        Ok(SshPrivateKey {
+            base_key,
+            public_key: SshPublicKey {
+                inner_key: base_public_key,
+                comment: String::new(),
+            },
+            passphrase,
+            kdf: Kdf {
+                name: kdf_name,
+                option: kdf_option,
+            },
+            _cipher_name: cipher_name,
+            check: check0,
+            comment,
+        })
+    }
+
+    fn encode(&self, mut stream: impl Write) -> Result<(), Self::Error> {
+        stream.write_all(AUTH_MAGIC.as_bytes())?;
+        stream.write_u8(b'\0')?;
+
+        if self.passphrase.is_some() {
+            stream.write_ssh_string(AES256_CTR)?;
+            stream.write_ssh_string(BCRYPT)?;
+
+            let salt = &self.kdf.option.salt;
+            let rounds = self.kdf.option.rounds;
```

Comment:
I'm not sure of the benefit of binding instead of using it directly?

Reply:
Readability?